### ISSUE TITLE
authCompletedTask

### ISSUE THREAD

**Issue by j3ffb** [Belicosus/AppAuth-Ionic#1]
Hi,

Thanks for sharing this, its been a great help. Could I ask you what the purpose of `authCompletedTask` is and how is it used? It doesn't seem to get set and `authCompletedResolve` seems to always be undefined when `requestAccessToken` is finished.

Thanks,
Jeff

**Comment by chougaard** [comment]
The purpose of the task is to allow us to wait for the asynchronous signin/token process to finish.
When we are done fetching the token, `authCompletedResolve()` gets called, which resolved `authCompletedTask`. This is very much related to your other question, where the `waitAuthenticated()` method is a wrapper around the task.

If you are working on implementing this in a simple app, please feel free to make a pull request. This repo could really do with a working proof of concept example in the form of a simple App.

**Comment by j3ffb** [comment]
Thanks for getting back to me. I can see `authCompletedResolve()` gets called, but it is always `undefined`, so an error is thrown when calling `requestAccessToken()`, I had to add `if(this.authCompletedResolve) this.authCompletedResolve();` as the last line, to get it to work. Is the caller responsible for setting it somewhere? Sorry, just trying to get my head around the flow.

**Comment by chougaard** [comment]
You shouldn't call `requestAccessToken()` from outside `AuthService`
you want to call `signin()` to start the process of signing in and retrieving tokens. When signin has finished, the callback of `startupAsync()` is called like this:
```
this.platform.ready().then(async () => {
    await this.authService.startupAsync(t => {
        //Register the callback for when authService has tried handling signin
        this.nav.setRoot(MainPage);
    });
});
```

**Comment by j3ffb** [comment]
Thanks for this. I still can't seem to figure out where `authCompletedResolve` gets set. In order to get `waitAuthenticated()` working I have to add this line `this.authCompletedResolve = _ => { };` before calling `this.authCompletedResolve();` at the end of `requestAccessToken()` and `requestWithRefreshToken()`.

**Comment by chougaard** [comment]
`authCompletedResolve` and `authCompletedReject` are callback methods for the Promise that is created in `resetAuthCompletedPromise()` [here](https://github.com/Belicosus/AppAuth-Ionic/blob/master/example/auth.service.ts#L85)

**Comment by j3ffb** [comment]
Ok I can see the problem now, as I was running this in the browser I had to make some changes to call `completeAuthorization()` as the page was reloaded by the redirect from the authentication request. This meant that the call to `requestAccessToken()` was called in a different session, so `resetAuthCompletedPromise()` was not called. I have added an extra call to `resetAuthCompletedPromise()` in `requestAccessToken()` and this is now working. Thanks for your help.
